src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,559 @@
+  }
+
+public:
+  bool try_to_build() {
+    // Try to create the .new file, we hold the global lock so there is no race
+    int fd = ::open(hm_lib_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(false);
+    }
+    ::close(fd);
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "make -B -C " << config.module_dir_ << " -f " << hm_mak_ << " 1>/dev/null";
+    // run the command synchronously, while holding the big_mutex
+    int rc = system(ss.str().c_str());
+    if (rc != 0) {
+      std::string base(config.module_dir_ + "/hm_" + key_);
+      fprintf(stderr, "makefile:\n%s", read_file_as_string(base+".mak").c_str());
+      fprintf(stderr, "errors:\n%s",   read_file_as_string(base+".err").c_str());
+    }
+    return (rc == 0);
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Master constructor - try to get module built in local file
+  config.ensure_module_dir_exists();
+  if (file_exists(lib_name_)) {
+    build_state_ = kPass;
+  } else {
+    // The file doesn't exist, let's build it
+    ModuleBuilder builder(options, source, include, key_);
+    build_state_ = (builder.try_to_build() ? kPass : kFail);
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  ssize_t binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  if (!file_exists(lib_name_)) {
+    // We hold big_mutex so there is no race
+    int fd = open(lib_name_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(0);
+    }
+    rc = write(fd, binary, binary_size);
+    assert(rc == binary_size);
+    ::close(fd);
+    ::chmod(lib_name_.c_str(), 0644);
+  }
+  build_state_ = kPass;
+  try_load_locked();
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+bool NativeModule::try_load_locked() {
+  if (load_state_ == kInit) {
+    if (is_global_) {
+      load_state_ = kPass;
+    } else {
+      auto handle = dlopen(lib_name_.c_str(), RTLD_GLOBAL|RTLD_NOW);
+      if (!handle) {
+        fprintf(stderr, "ERROR: dlopen failed: %s\n", dlerror());
+      }
+      load_state_ = (handle ? kPass : kFail);
+      if (handle) dlopen_handle_ = handle;
+    }
+  }
+  return (load_state_ == kPass);
+}
+
+std::vector<char> NativeModule::get_binary() {
+  std::lock_guard<std::mutex> mylock(big_mutex);
```

Comment:
I don't think the lock is necessary?

Reply:
The big_mutex is now the only protection against reading a zero-length or partially-written lib file,
so I think it is necessary here.

Reply:
Two comments:

On reflection, it seems get_binary should test build_status_, no?  If the build failed or wasn't written, we shouldn't be trying to load the file.

Second, the NativeModule construction and build/write is synchronous, so if we're here, in a NativeModule method, it must have built successfully, no?

Reply:
The build must have happened, but it might have build_state_ == kFail

Reply:
Right, so you should check build_state_, but why the lock?  Anyway, doesn't hurt.

Reply:
I see your point about how we shouldn't need big_mutex here because
a) the 0-or-1 NativeModule-per-key rule means that the NativeModule being called with get_binary must be the only NativeModule dealing with this key
and
b) since we're in a method, this Native Module must have run its constructor, and thus it
is *not* still writing lib_file

I'm still inclined to hold the big_mutex here, to make the mutual-exclusion explicit and obvious.

Reply:
> I'm still inclined to hold the big_mutex here, to make the mutual-exclusion explicit and obvious.

Great, no object to explicit and obvious!